### PR TITLE
Bug fix - Threading issues

### PR DESCRIPTION
**Pull request by Thurman1776-**
## Description 
Turns out scanning large projects is pushing SwiftUI to its limits. 

### Tech details
I've profiled the app with Instruments before & after the current changes & the issue happens in both cases, just in 2 different scenarios. 

1. Current behaviour - First launch works fine, no main thread bottleneck. However, as soon as you start filtering dependencies in the search field, the main thread gets clogged. 
2. Behaviour with these changes - First launch and subsequent filters are consistently clogged. This is obviously worse than the old/current behaviour but at least it is easier to reason about the problem (at least I hope), whereas before it just happens by pure luck the first launch isn't flooding the main thread. 

The culprit is a C++ function with the following signature: 
`pair<(anonymous namespace)::TupleCacheEntry*, swift::MetadataResponse> swift::LockingConcurrentMap<(anonymous namespace)::TupleCacheEntry, (anonymous namespace)::TupleCacheStorage>::getOrInsert<(anonymous namespace)::TupleCacheEntry::Key, swift::MetadataRequest&, swift::TargetValueWitnessTable<swift::InProcess> const*&>((anonymous namespace)::TupleCacheEntry::Key, swift::MetadataRequest&, swift::TargetValueWitnessTable<swift::InProcess> const*&)`

which is internally calling [swift_getTupleTypeMetadata2](https://github.com/apple/swift/blob/main/stdlib/public/runtime/Metadata.cpp#L1966-L1974) which then calls multiple `libSwiftCore` APIs as well as `SwiftUI` ones. 

I could not find much references out there on the topic, but as this seems to work fine with relatively small projects/frameworks, I think there's little we can do for now. 
Perhaps after updating to latest version of Swift some of these issues have been somewhat fixed (could be that these performance bottlenecks are influenced by the fact that we are [type erasing](https://github.com/Thurman1776-/Uncomplect-us/blob/master/Mac-App/Frontend/Frontend/Views/ExpandableList/DependencyListView.swift#L26-L36) concrete SwiftUI views, which is a known issue).

### CODE REVIEW

File: Mac-App/Mac-App/Transformers/NavigationTransformer.swift [hunk 1]
```diff
@@ -32,7 +32,7 @@ final class NavigationTransformer: StateTransforming, StateRepresentableViewInpu
     }
 
     func startListening() {
-        cancellable = stateObserver.$currentState.sink { [unowned self] in self.emitNewState($0) }
+        cancellable = stateObserver.$currentState.receive(on: DispatchQueue.main).sink { [unowned self] in self.emitNewState($0) }
```
Comment (Thurman1776-) [inline Mac-App/Mac-App/Transformers/NavigationTransformer.swift:35]: This transformer takes care of views transition, thus we need to execute state changes in the main thread.

File: Mac-App/Mac-App/BackendSubscribers/BackendSubscription.swift [hunk 2]
```diff
@@ -28,18 +29,32 @@ struct BackendSubscription: StateSubscription {
...
         let navigationTransformer = NavigationTransformer(
             window: window,
             listViewInput: listViewTransformer.viewInput,
             projectDetailsViewInput: projectDetailsTransformer.viewInput
         )
-        BackendAPI.subscribe(navigationTransformer.stateObserver) { $0.select(NavigationData.State.init) }
+        BackendAPI.subscribe(navigationTransformer.stateObserver) { appState -> Subscription<Frontend.NavigationData.State> in
+            let skippingOnSameSubState = appState.skip { $0.navigationState == $1.navigationState }
+            return skippingOnSameSubState.select(NavigationData.State.init)
+        }
```
Comment (Thurman1776-) [inline Mac-App/Mac-App/BackendSubscribers/BackendSubscription.swift:60]: All of these validations ensure unnecessary triggers to state observers.

**Comment by Thurman1776-** [comment]
@emrepun 

I fixed the issue completely! 🎉  

The fix was split in 2 commits: 

f023482bb3c410a1b64c7daa9b7c03ae2ac91ffa - Prevents multiple searches to be executed. That's a nasty side effect I realised when [switching window content view](https://github.com/Thurman1776-/Uncomplect-us/blob/master/Mac-App/Mac-App/Transformers/NavigationTransformer.swift#L56-L60). In other words, when the app transitions from `InputView` to `MainSplitView` the system calls again the [onCommit](https://github.com/Thurman1776-/Uncomplect-us/blob/master/Mac-App/Frontend/Frontend/Views/InputView/InputView.swift#L22-L23) closure, meaning searching effectively twice! 😱  

5379e0dc5962cdf1dd68bfd120890b9e92a4d28c - This one took a bit more of investigation in Instruments. I kept digging on the stack trace & found `CollectionChanges.formChanges` being the concrete SwiftUI culprit. A quick Google search pointed me to the fix (the link of the solution is included in the commit) 

This means scanning large projects & filtering results is no longer a bottleneck! 👯 
Screenshot 2020-10-30 at 21 11 57

File: Mac-App/Backend/Backend/BackendAPI/Middleware/DependencyGraphSideEffects/DependencyGraphSideEffects.swift [hunk 3]
```diff
@@ -18,7 +18,7 @@ func dependencyGraphSideEffects(parser: @escaping SwiftDepsParserType) -> Middle
 
                 switch action {
                 case let DependencyGraphAction.mapFrom(swiftDeps):
-                    dispatchAsyncOnBackendQueue {
+                    dispatchAsyncOnConcurrentBackendQueue {
```
Comment (emrepun) [inline Mac-App/Backend/Backend/BackendAPI/Middleware/DependencyGraphSideEffects/DependencyGraphSideEffects.swift]: is Backend really needed in the name of the method? Or can it be replaced by Background?

File: Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    ) {
+        if let _ = DispatchQueue.getSpecific(key: key) {
+            work()
+        } else {
+            async {
+                work()
+            }
+        }
+    }
+}
```
Comment (emrepun) [inline Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift]: Can we rename the method so it is obvious what are we really checking? Or a comment perhaps?

**Comment by emrepun** [comment]
> @emrepun
> 
> I fixed the issue completely! 🎉
> 
> The fix was split in 2 commits:
> 
> [f023482](https://github.com/Thurman1776-/Uncomplect-us/commit/f023482bb3c410a1b64c7daa9b7c03ae2ac91ffa) - Prevents multiple searches to be executed. That's a nasty side effect I realised when [switching window content view](https://github.com/Thurman1776-/Uncomplect-us/blob/master/Mac-App/Mac-App/Transformers/NavigationTransformer.swift#L56-L60). In other words, when the app transitions from `InputView` to `MainSplitView` the system calls again the [onCommit](https://github.com/Thurman1776-/Uncomplect-us/blob/master/Mac-App/Frontend/Frontend/Views/InputView/InputView.swift#L22-L23) closure, meaning searching effectively twice! 😱
> 
> [5379e0d](https://github.com/Thurman1776-/Uncomplect-us/commit/5379e0dc5962cdf1dd68bfd120890b9e92a4d28c) - This one took a bit more of investigation in Instruments. I kept digging on the stack trace & found `CollectionChanges.formChanges` being the concrete SwiftUI culprit. A quick Google search pointed me to the fix (the link of the solution is included in the commit)
> 
> This means scanning large projects & filtering results is no longer a bottleneck! 👯
> Screenshot 2020-10-30 at 21 11 57

Wow when I first saw setting input string to empty commit, I didn't know it was a fixing commit actually 😅 Also the link in the commit is very helpful to understand. 🎉

File: Mac-App/Backend/Backend/BackendAPI/Middleware/DependencyGraphSideEffects/DependencyGraphSideEffects.swift (hunk 3 above)
Comment (Thurman1776-) [reply Mac-App/Backend/Backend/BackendAPI/Middleware/DependencyGraphSideEffects/DependencyGraphSideEffects.swift]: Hmm you're right, it's already in the `Backend` framework so it is implicit. Also, this function is not even available outside the module. Will rename 👍

File: Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift (hunk 4 above)
Comment (Thurman1776-) [reply Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift]: Yeah, I was struggled naming this. Do you have any suggestions?

File: Mac-App/Backend/Backend/BackendAPI/Middleware/DependencyGraphSideEffects/DependencyGraphSideEffects.swift (hunk 3 above)
Comment (Thurman1776-) [reply Mac-App/Backend/Backend/BackendAPI/Middleware/DependencyGraphSideEffects/DependencyGraphSideEffects.swift]: https://github.com/Thurman1776-/Uncomplect-us/pull/28/commits/16f7a12d6e39e5782de97cb86a902df2f38af5c4

File: Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift (hunk 4 above)
Comment (Thurman1776-) [reply Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift]: What do you think of this? 
https://github.com/Thurman1776-/Uncomplect-us/pull/28/commits/14d827f302f8514b3cc70468566d5b4d8031b52a

File: Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift [hunk 5]
```diff
@@ -7,8 +7,14 @@
...
-    func asyncWithCheck(
-        key: DispatchSpecificKey<String>,
+    /// This method checks current Queue's context using its key before executing the closure to prevent
+    /// unnecessary thread hops
+    /// - Parameters:
+    ///   - key: Key assigned to a Queue
+    ///   - work: Closure to be run in a thread that matches the receiver's key.
+    ///   Else it is enqueued to a different Queue
+    func asyncCheckingCurrentQueue(
+        for key: DispatchSpecificKey<String>,
```
Comment (emrepun) [inline Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift:17]: Nice! 🎉  @Thurman1776-

File: Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift (hunk 4 above)
Comment (emrepun) [reply Mac-App/Backend/Backend/Shared/Extensions/DispatchQueue.swift]: Nice one @Thurman1776-

**Review by emrepun — APPROVED** [review]
🎉

**Pull request merged**